File: Exercism/csharp/protein-translation/ProteinTranslation.cs
using System;
using System.Collections.Generic;
using System.Linq;

public static class ProteinTranslation
{
    private static string Match(string strand) => strand.Substring(0, 3) switch
    {
        "AUG" => "Methionine",
        "UUU" or "UUC" => "Phenylalanine",
        "UUA" or "UUG" => "Leucine",
        "UCU" or "UCC" or "UCA" or "UCG" => "Serine",
        "UAU" or "UAC" => "Tyrosine",
        "UGU" or "UGC" => "Cysteine",
        "UGG" => "Tryptophan",
        "UAA" or "UAG" or "UGA" => "STOP",
        var s => throw new ArgumentException($"Invalid strand: {s}")
    };

    private static string[] ProcessProteins(string strand, IEnumerable<string> processed) => 
      strand?.Length < 3 || Match(strand) == "STOP"
        ? processed.ToArray()
        : ProcessProteins(strand.Substring(3), processed.Append(Match(strand)));

    public static string[] Proteins(string strand) => ProcessProteins(strand, new List<string>());
}

File: Exercism/fsharp/protein-translation/ProteinTranslation.fs
module ProteinTranslation

let codonToProtein = function
    | "AUG" -> "Methionine"
    | "UUU" | "UUC" -> "Phenylalanine"
    | "UUA" | "UUG" -> "Leucine"
    | "UCU" | "UCC" | "UCA" | "UCG" -> "Serine"
    | "UAU" | "UAC" -> "Tyrosine"
    | "UGU" | "UGC" -> "Cysteine"
    | "UGG" -> "Tryptophan"
    | "UAA" | "UAG" | "UGA" -> "Stop"
    | n -> failwith $"Invalid codon '{n}' supplied"

let proteins rna =
    rna
    |> Seq.chunkBySize 3
    |> Seq.map (System.String >> codonToProtein)
    |> Seq.takeWhile ((<>) "Stop")
    |> List.ofSeq
